### DISCUSSION TITLE
Process hangs around after closing last window with 'Destroying EGL context'

### DISCUSSION THREAD

**Discussion by jorismak** [ghostty-org/ghostty#3495] [Q&A]
Linux, Ubuntu 24.04, Wayland.

I'm starting ghostty from another terminal now to set things up, and I notice that the last line in the terminal is 'Destroying EGL context', and then the ghostty process just hangs there.

I've changed the `quit-after-last-window-closed` option to `true` in the config to make sure it quits, but it still stays around.

So, I have a global keybind on my system that starts a terminal (ctrl-alt-T), and then when I close the process, it stays around without windows. So every time I start a new terminal to get a window, I spawn a new process that stays around forever.

I fixed that now by forcing `gtk-single-instance` mode to `true`, that way no new processes will be spawned.
But the ghostty process not being able to exit and requiring a CTRL-C / break seems weird.

**Comment by isgj** [comment]
To add something here, also when you open it from the menu a similar behavior happens.
When you click the close window button, it asks that all the sessions will be terminated, after confirming (clicking yes), the window is closed and the process is terminated.

If you press `CTRL-D` (which I happen to use a lot), the window will be closed but the process will hang.

**Comment by jorismak** [comment]
Ha, you're right.
(Disabled gtk-single-instance for testing)

Closing the window by clicking the  X (as in, closing through the window manager) exits with:

```
Destroying EGL context
Destroying EGL context
info(io_exec): read thread got quit signal
info(surface): surface closed addr=24c85c30
info(sentry): sentry envelope does not contain crash, discarding
```

...and really exits the ghostty process.

Pressing CTRL+D in the last window (or just typing `exit`) exits with:

```
Destroying EGL context
```

...and then just hangs there.

**Comment by lokesh-krishna** [comment]
I have .desktop entries to launch certain programs in ghostty with `ghostty -e <app-name>`. Whenever I launch one of these, from an existing terminal or a launcher, using my window manager's quit keybinding just has it hanging. If I launch it from a terminal then the message it hangs on is:
```
Destroying EGL context
Destroying EGL context
```

**Comment by lokesh-krishna** [comment]
Logs when launching senpai with `ghostty -e senpai` and quitting using the `/quit` command in senpai: [log](https://paste.low-key.me/low-key/4f993a63378f44dcad0a0f790e4c7b3b)

Logs when quitting using my window manager's keybinding or ghostty's quit action with a keybinding: [log](https://paste.low-key.me/low-key/4f993a63378f44dcad0a0f790e4c7b3b)

**Comment by lokesh-krishna** [comment]
Has anyone found any workaround to this? I'm confused about whether this is not an issue because the problem hasn't been described well enough or if this is somehow expected behaviour.

**Reply by morph027** [reply]
Nope, nothing, but it happens here too and i don't feel it's supposed to be this way.

**Reply by jorismak** [reply]
Sort of the same on my end. This feels as a bug and not 'a discussion'. But i don't want to make a point about it during the release and the massive wall of issues coming their way. As long as it is known, eventually 😉.

So the problem is that ghostty won't quit when the terminal closes by itself. Ghostty quits ok when the terminal window is closed by the window manager.

For my use case, I force enabled the 'gtk single instance' as a workaround. That means ghostty will hang around , but when I launch a new ghostty process (by my global hotkey) it will be the older 'hanging' process that creates a new window , and the newly launched ghostty process exits immediately.
So, I don't have a bunch of hanging ghostty processes, just the one.

This also means the problem we're encountering isn't really 'hanging'. Ghostty will not exit and sticks around, but apparently listens and functions fine on the gtk IPC signals (whatever ghostty uses to communicate between processes ). 

But gtk single instance disabled just doesn't exit and that's wrong. And since the default is 'auto' and it often decides NOT to use single instance mode, there is a wold growth of ghostty processes.

As I said , in _my_ case I can _work around_ it, by enabling gtk single instance. Leaving it disabled or on the default auto is causing issues.

**Reply by lokesh-krishna** [reply]
Using the `-e` flag seems to automatically disable `gtk-single-instance` so not even sure how to workaround it :(

**Reply by morph027** [reply]
I agree....this is sort of a "bug", but not critical one. I mean, normally i do not close the terminal anyway, because of....you knowm work to be done :joy:

**Reply by basile-henry** [reply]
This is a bug that I have been encountering as well. I thought it was more Ctrl-D related, and I hadn't seen this discussion so I contributed to this discussion instead: https://github.com/ghostty-org/ghostty/discussions/3596

> I can work around it, by enabling gtk single instance

I ended up figuring this out as well, and since it doesn't leak processes as much this way it's fine as a workaround, but I think this needs a proper fix.

I'm not entirely sure these discussions work as well as normal issues :sweat_smile: 

> As long as it is known, eventually

I hope so :crossed_fingers:

**Reply by pgmac** [reply]
I've been experiencing this same behaviour, too. IE:

- "hanging" when using CTRL-D or typing "exit".
- Exiting normally when clicking the "X" on the window title bar.

I tried using the GTK single instance, but that behaved the same for me and doesn't really work for my use case. I create and exit many terminals throughout (to execute commands and/or SSH connections).

This is on my Ubuntu 24.04 LTS laptop. 
I've compiled the binary (using docker) in Ubuntu 24.04 and 24.10. Aside from different versions of GTK at runtime, both behave the same.

I also compiled using a RHEL docker image, as another person I know was able to use this on their (Asahi Linux) system.

Right now, I suspect this might be an Ubuntu OpenGL implementation/integration thing. But I don't know how to further investigate/debug this deeper.

I'm also using Gnome Shell.  I haven't tried any other window managers. I don't know if that would make any difference, but it might be worth looking into. Just trying to rule out any other variables.

**Reply by jorismak** [reply]
That threat at least seems to indicate it's not a Linux only issue, which helps in my mind.

Why these issues get 'demoted' to discussions (this is how it feels to me at least ) I don't know, but as I said, I'm not going to limit my experience 😉.

**Reply by cespare** [reply]
This reproduces 100% of the time for me as well. Also Ubuntu 24.04 and wayland.

AFAICT this should be an issue, not a discussion, according to the guidelines outlined in https://github.com/ghostty-org/ghostty/blob/main/CONTRIBUTING.md, since we have very straightforward repro steps.

**Comment by goonzoid** [comment]
It's possible that this _might_ be fixed by https://github.com/ghostty-org/ghostty/pull/4605, but I don't have easy access to a Linux machine to test on right now. If someone feels like checking out my fork and giving it a try, that might be illuminating.

**Reply by slovdahl** [reply]
Doesn't seem to work very well on at least Ubuntu 24.04 right now: https://github.com/ghostty-org/ghostty/pull/4605#issuecomment-2571516201

**Reply by goonzoid** [reply]
I'm now pretty convinced that this is a different issue than the one #4605 fixes, fwiw.

**Comment by bkw777** [comment]
I think I have the same problem. ghostty 1.1.0 from the ubuntu ppa
I wanted to add ghostty as an option to https://github.com/bkw777/mainline
I have built-in preconfigured support for almost 20 different terminals, and most of them take an -e argument for a command to run like xterm -e, and they all work fine but `ghostty -e` hangs in some way on exit. The binary being run inside the terminal exits, and the terminal window goes away, but the parent process is still frozen as if the child terminal process were still running. Which it is:

While the child process is running (waiting for an \[Enter\] to exit itself, not using wait-after-command:

```
bkw@fw:~/src/mainline$ ps -ef |egrep '(ghost|mainline)'
bkw       107923    2007  0 12:56 ?        00:00:00 mainline-gtk
bkw       107956  107923  0 12:56 ?        00:00:00 sh -c -- ghostty -e mainline --from-gui --pause install 6.12.2
bkw       107957  107956  0 12:56 ?        00:00:01 ghostty -e mainline --from-gui --pause install 6.12.2
bkw       107984  107957  0 12:56 pts/4    00:00:00 /bin/sh -c mainline --from-gui --pause install 6.12.2
bkw       107986  107984  0 12:56 pts/4    00:00:00 mainline --from-gui --pause install 6.12.2
bkw       117979    2673  0 14:32 pts/0    00:00:00 grep -E --color=auto (ghost|mainline)
bkw@fw:~/src/mainline$ 
```

After the child process exit()  
The ghostty window went away, and the executable that was run in the terminal is also gone.
The `mainline --from-gui ...` process is gone, but `ghosttty -e ...` is still there.
```
bkw       107923    2007  0 12:56 ?        00:00:00 mainline-gtk
bkw       107956  107923  0 12:56 ?        00:00:00 sh -c -- ghostty -e mainline --from-gui --pause install 6.12.2
bkw       107957  107956  0 12:56 ?        00:00:01 ghostty -e mainline --from-gui --pause install 6.12.2
bkw       118301    2673  0 14:37 pts/0    00:00:00 grep -E --color=auto (ghost|mainline)
bkw@fw:~/src/mainline$ 
```

The parent mainline-gtk process is hung waiting for ghostty to exit.

After letting the window manager kill the parent window, ghostty and the shell that ran it still there, now with parent tranferred to pid 1.
```
bkw@fw:~/src/mainline$ ps -ef |egrep '(ghost|mainline)'
bkw       107956       1  0 12:56 ?        00:00:00 sh -c -- ghostty -e mainline --from-gui --pause install 6.12.2
bkw       107957  107956  0 12:56 ?        00:00:01 ghostty -e mainline --from-gui --pause install 6.12.2
bkw       118438    2673  0 14:39 pts/0    00:00:00 grep -E --color=auto (ghost|mainline)
bkw@fw:~/src/mainline$ 
```

**Comment by cespare** [comment]
This is fixed, at least for me, by https://github.com/ghostty-org/ghostty/pull/6796.

(It was reproducible 100% of the time before.)

**Comment by jorismak** [comment]
Cool, will try later!
